operator/internal/decommissioning/statefulset_decomissioner.go

```diff
@@ -0,0 +1,646 @@
+		currentlyDecommissioningBrokers = append(currentlyDecommissioningBrokers, downedNode)
+	}
+
+	sortBrokers := func(set []int) []int {
+		// sort by simple node id
+		sort.SliceStable(set, func(i, j int) bool {
+			return set[i] < set[j]
+		})
+		return set
+	}
+
+	healthyBrokers := sortBrokers(healthyNodes.Values())
+	brokersToDecommission = sortBrokers(brokersToDecommission)
+	brokersToIgnore = sortBrokers(brokersToIgnore)
+	currentlyDecommissioningBrokers = sortBrokers(currentlyDecommissioningBrokers)
+
+	formatBrokerList := func(set []int) string {
+		return strings.Join(functional.MapFn(strconv.Itoa, set), ", ")
+	}
+
+	log.V(traceLevel).Info(fmt.Sprintf(
+		"healthy brokers: [%s], ignored: [%s], to decommission: [%s], decommissioning: [%s]",
+		formatBrokerList(healthyBrokers),
+		formatBrokerList(brokersToIgnore),
+		formatBrokerList(brokersToDecommission),
+		formatBrokerList(currentlyDecommissioningBrokers),
+	))
+
+	if len(currentlyDecommissioningBrokers) != 0 {
+		// we skip decommissioning our next broker since we already have some node decommissioning in progress
+		return true, nil
+	}
+
+	if len(brokersToDecommission) > 0 {
+		// only record the event here since this is when we trigger a decommission
+		s.recorder.Eventf(set, corev1.EventTypeNormal, eventReasonBroker, "brokers needing decommissioning: [%s], decommissioning: %d", formatBrokerList(brokersToDecommission), brokersToDecommission[0])
+
+		if err := adminClient.DecommissionBroker(ctx, brokersToDecommission[0]); err != nil {
+			log.Error(err, "decommissioning broker", "broker", brokersToDecommission[0])
+			return false, err
+		}
+	}
+
+	// we should have decommissioned something above, so requeue and wait for it to finish
+	return true, nil
+}
+
+// findUnboundVolumeClaims fetches any PVCs associated with the StatefulSet that aren't actively attached
+// to a pod.
+//
+// Much of this code is copied from the original decommissioner and refactored, but the basic idea
+// is:
+//
+// 1. Pull any pods matching the labels for the stateful set's pod template that are in the same namespace
+// 2. Pull any pvcs matching the labels for the stateful set's volume claim template (though the component adds a "NAME-statefulset")
+// 3. Find unbound volumes by checking that the pods we pulled reference every volume claim
+//
+// NB: this follow the original implementation that has a potential race-condition in the cache, where a PVC may come online and be in-cache
+// but the corresponding pod has not yet populated into the cache. In this case the PVC could be marked for deletion
+// despite the fact that it's still bound to a pod. In such a case the pvc-protection finalizer put in-place by core keeps the
+// PVC from being deleted until the pod is deleted. Due to the skip of already-deleted PVCs below, these PVCs should
+// just get GC'd when the pod is finally decommissioned.
+//
+// TODO: will this cause issues if a PVC is deleted and before it's GC'd a Pod comes up?
```

Comment:
One thing to note about this `TODO` -- I may have run into this on one test run while I was writing the tests. It looked like a pod may have gotten stuck because its PVC was getting deleted, while it was attempting to be provisioned. If we want to be a bit more guarded here I could always just add a minimum age before we delete an unbound PVC and more aggressively re-reconcile the state of the world through turning down the cache resync period? Not sure if that'll handle everything either though?

Either that or I could do something similar to the ordinal collision code for the pods?

1. Grab all PVCs -- map back to ordinals (they're all named `datadir-NAME-ORDINAL`)
2. see if any are above the max statefulset ordinal, if so _and_ unbound, delete
3. if the PVC is unbound, but not above the max ordinal range, keep it around